### PR TITLE
Add support for CPU

### PR DESCRIPTION
**Pull request by Tabrizian**
```
Models (Inference):                                                                                                                                                                                                                                       
Model                       Batch   Concurrency   Model Config Path               Instance Group   Preferred Batch Sizes   Satisfies Constraints                                                                                                          
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i6    1/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i6    1/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i6    1/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i7    2/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i7    2/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i7    2/CPU            [2]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i9    2/CPU            [4]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i9    2/CPU            [4]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i9    2/CPU            [4]                     Yes                                                                                                                            
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i10   1/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i10   1/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i10   1/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i13   2/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i13   2/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i13   2/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i8    1/CPU            [4]                     Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i8    1/CPU            [4]                     Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i8    1/CPU            [4]                     Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i0    1/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i0    1/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i0    1/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i2    1/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i2    1/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i2    1/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i5    2/CPU            [1]                     Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i5    2/CPU            [1]                     Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i5    2/CPU            [1]                     Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i12   1/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i12   1/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i12   1/CPU            [16]                    Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i11   2/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i11   2/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i11   2/CPU            [8]                     Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i3    2/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i3    2/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i3    2/CPU            [128]                   Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i1    2/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i1    2/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i1    2/CPU            Disabled                Yes
resnet_v1_50_cpu_graphdef   1       4             resnet_v1_50_cpu_graphdef_i4    1/CPU            [1]                     Yes
resnet_v1_50_cpu_graphdef   1       2             resnet_v1_50_cpu_graphdef_i4    1/CPU            [1]                     Yes
resnet_v1_50_cpu_graphdef   1       1             resnet_v1_50_cpu_graphdef_i4    1/CPU            [1]                     Yes
```

The new output will contain a single table.

[result_summary.pdf](https://github.com/triton-inference-server/model_analyzer/files/6387680/result_summary.pdf)

### CODE REVIEW

File: model_analyzer/plots/plot.py [hunk 1]
```diff
@@ -158,4 +159,4 @@ def save(self, filepath):
             this plot should be saved to
         """
 
-        self._fig.savefig(os.path.join(filepath, self._name))
+        self._fig.savefig(os.path.join(filepath, self._name), dpi=100, bbox_inches='tight')
```
Comment (aramesh7) [inline model_analyzer/plots/plot.py:162]: What exactly is this doing?

File: model_analyzer/record/metrics_manager.py [hunk 2]
```diff
@@ -130,35 +136,47 @@ def profile_model(self, perf_config, perf_output_writer=None):
         model_non_gpu_metrics = list(perf_analyzer_metrics.values()) + list(
             model_cpu_metrics.values())
 
-        return model_gpu_metrics, model_non_gpu_metrics
+        measurement = None
+        if model_gpu_metrics is not None and model_non_gpu_metrics is not None:
```
Comment (aramesh7) [inline model_analyzer/record/metrics_manager.py:140]: This used to be here, but then we moved it to `_execute_run_config` for the run config search PR. Wondering what made you move it back here? I agree it looks cleaner.

File: model_analyzer/reports/report_manager.py [hunk 3]
```diff
@@ -231,6 +245,51 @@ def _build_summary_table(self, report_key, num_measurements, gpu_name):
             summary_table.insert_row_by_index(row)
         return summary_table, summary_sentence
 
+    def _build_summary_cpu_table(self, report_key, num_measurements):
+        """
```
Comment (aramesh7) [inline model_analyzer/reports/report_manager.py:249]: This function seems to have a lot in common with `_build_summary_table`. Would it be possible to simply add an extra argument like `cpu_only` or soemthing to account for the modified functionality?

File: model_analyzer/result/result_manager.py [hunk 4]
```diff
@@ -154,23 +137,44 @@ def create_tables(self, gpu_specific_metrics, non_gpu_specific_metrics):
...
-            else:
-                raise TritonModelAnalyzerException(
-                    f'GPU output field "{gpu_output_field}", does not exist')
-        self._gpu_output_fields = gpu_output_fields
-
-        # Model GPU Metrics
-        self._add_result_table(table_key=self.model_gpu_table_key,
-                               title='Models (GPU Metrics)',
-                               headers=gpu_output_headers)
+        if not self._is_cpu_only:
```
Comment (aramesh7) [inline model_analyzer/result/result_manager.py:140]: This function is starting to get bloated. Perhaps we should split it up into `_create_server_table`, `_create_inference_table`, and `_create_gpu_table`.

**Review by aramesh7 — CHANGES_REQUESTED** [review]

File: model_analyzer/plots/plot.py (hunk 1 above)
Comment (Tabrizian) [reply model_analyzer/plots/plot.py:162]: It is a 'trick' that I remember from grad school days to make the plots look good. Second argument make sure that the subplot fits the figure and the other argument increases the quality of image.

File: model_analyzer/record/metrics_manager.py (hunk 2 above)
Comment (Tabrizian) [reply model_analyzer/record/metrics_manager.py:140]: Just wanted to make sure that it returns a measurement similar to `profile_server`.

File: model_analyzer/model_manager.py [hunk 5]
```diff
@@ -44,13 +45,15 @@ def __init__(self, config, client, server, metrics_manager,
...
-        self._run_search = RunSearch(config=config)
+        self._is_cpu_only = config.cpu_only
+        self._run_search = RunSearch(config=config, cpu_only=self._is_cpu_only)
         self._last_config_variant = None
         self._run_config_generator = RunConfigGenerator(config=config,
                                                         client=self._client)
 
         # Generate the output model repository path folder.
         self._output_model_repo_path = config.output_model_repository_path
+        print(config.output_model_repository_path)
```
[the remark this answers is not in the document]
Comment (Tabrizian) [reply model_analyzer/model_manager.py:56]: Rogjt. good catch.

File: model_analyzer/reports/report_manager.py (hunk 3 above)
Comment (Tabrizian) [reply model_analyzer/reports/report_manager.py:249]: Sure, will fix it in the review.

File: model_analyzer/record/metrics_manager.py [hunk 6]
```diff
@@ -28,7 +29,7 @@ class MetricsManager:
     This class handles the profiling
     categorization of metrics
     """
-    def __init__(self, config, metric_tags, server, result_manager):
+    def __init__(self, config, metric_tags, server, result_manager, gpus):
```
Comment (aramesh7) [inline model_analyzer/record/metrics_manager.py:32]: Is this argument necessary still? We're passing in `config`. So could we use the `GPUDeviceFactory.get_analyzer_gpus` function call within this constructor?

**Review by aramesh7 — CHANGES_REQUESTED** [review]

**Review by aramesh7 — DISMISSED** [review]

**Review by aramesh7 — APPROVED** [review]

**Pull request merged**